CRM/Utils/Check/Component/Case.php

```diff
@@ -397,4 +397,114 @@ public function checkRelationshipTypeProblems() {
+          '<p>' . ts('Unable to locate xml file for Case Type "<em>%1</em>".',
+          [
+            1 => htmlspecialchars(empty($case_type['title']) ? $dao->id : $case_type['title']),
+          ]) . '</p>',
+          ts('Missing Case Type Definition'),
+          \Psr\Log\LogLevel::ERROR,
+          'fa-exclamation'
+        );
+        continue;
+      }
+
+      if (empty($case_type['definition']['caseRoles'])) {
+        $messages[] = new CRM_Utils_Check_Message(
+          __FUNCTION__ . "missingcaseroles",
+          '<p>' . ts('CaseRoles seems to be missing in the xml file for Case Type "<em>%1</em>".',
+          [
+            1 => htmlspecialchars(empty($case_type['title']) ? $dao->id : $case_type['title']),
+          ]) . '</p>',
+          ts('Missing Case Roles'),
+          \Psr\Log\LogLevel::ERROR,
+          'fa-exclamation'
+        );
+        continue;
+      }
+
+      // Loop thru each role in the xml.
+      foreach ($case_type['definition']['caseRoles'] as $role) {
+        $name_to_suggest = NULL;
+        $xml_name = $role['name'];
+        if (isset($names_a_b[$xml_name]) || isset($names_b_a[$xml_name])) {
+          // It matches a name, so either name and label are the same or it's
+          // an edge case already dealt with by core status checks, so do
+          // nothing.
+          continue;
+        }
+        elseif (isset($labels_b_a[$xml_name])) {
+          // $labels_b_a[$xml_name] gives us the id, so then look up name_b_a
+          // from the original relationship_types array which is keyed on id.
+          // We do b_a first because it's the more standard one, although it
+          // will only make a difference in edge cases which we leave to the
+          // other checks.
+          $name_to_suggest = $relationship_types[$labels_b_a[$xml_name]]['name_b_a'];
+        }
+        elseif (isset($labels_a_b[$xml_name])) {
+          $name_to_suggest = $relationship_types[$labels_a_b[$xml_name]]['name_a_b'];
+        }
+
+        // If it didn't match any name or label then that's weird.
+        if (empty($name_to_suggest)) {
+          $messages[] = new CRM_Utils_Check_Message(
+            __FUNCTION__ . "invalidcaserole",
+            '<p>' . ts('CaseRole "<em>%1</em>" in the xml file for Case Type "<em>%2</em>" doesn\'t seem to match any existing relationship type.',
+            [
+              1 => htmlspecialchars($xml_name),
+              2 => htmlspecialchars(empty($case_type['title']) ? $dao->id : $case_type['title']),
+            ]) . '</p>',
+            ts('Invalid Case Role'),
+            \Psr\Log\LogLevel::ERROR,
+            'fa-exclamation'
+          );
+        }
+        else {
+          $messages[] = new CRM_Utils_Check_Message(
+            __FUNCTION__ . "suggestedchange",
+            '<p>' . ts('Suggest changing Case Role "<em>%1</em>" in the xml file for Case Type "<em>%2</em>" to "<em>%3</em>". Using label is deprecated as of version 5.20.',
```

Comment:
We should be a little more directive and be a little clearer what the proposed change.  What do you think about this?
```suggestion
            '<p>' . ts('Please edit the XML file for case type "<em>%2</em>" so that the case role label "<em>%1</em>" is changed to its corresponding name "<em>%3</em>" (assuming a client has a "%1" relationship to the contact in the case role). Using label is deprecated as of version 5.20.',
```

Also I added a parenthetical because someone with old XML files might not be thinking as specifically about the direction of the relationship when they set it up.  I don't want to get into what needs to happen if they have it messed up, but at least it will be clear that they have a more significant problem.

Reply:
Thanks for the review. I'd like to leave out the parenthetical. I realize now my screenshot above might be misleading because it is the reverse relationship from what you would typically see in an xml file, but the status check will suggest the correct direction depending on what they have in the xml already, subject to edge cases which are dealt with elsewhere.

Reply:
First, I messed up the proposed parenthetical; it should be reversed:
```
(assuming the contact in the case role has a "%1" relationship to the client)
```

But the reason I wanted the extra context is that back in [dev/core#530](https://lab.civicrm.org/dev/core/issues/530) / #13916 we made the direction matter.  Before then, the code generally assumed that the client was on the A side of the relationship, and the labels were used willy-nilly.

My concern is that the typical case type writer might have had a similar attitude about the case role.  In the following situation (getting one last table in!)

| Direction | Label          | Name           |
| --------- | -------------- | -------------- |
| `a_b`     | Grandchild of  | Grandson of |
| `b_a`     | Grandparent of | Grandmother of |

someone could potentially have listed a case role as the A-to-B `Grandchild of` rather than `Grandparent of` if the client is the A because much of the UI would say "Grandchild of" when referring to the role.

Now that we have to take direction into account, if the XML says `Grandchild of`, that implies that the client is the B, and the system check will tell someone to change it to `Grandson of`.  They really should be setting it to `Grandmother of`.  We can't get into all of this in the text of a system check, but we can at least flag that the advice is making an assumption:

> Please edit the XML file for case type "*Children's services*" so that the case role label "*Grandchild of*" is changed to its corresponding name "*Grandson of*" (assuming the contact in the case role has a "Grandchild of" relationship to the client). Using label is deprecated as of version 5.20.

The hope is that someone would see the assumption, realize that it's wrong, and dig in some more instead of directly applying the proposed change.

Reply:
Ah yes tables. I see what you're saying now, that there could be sort of a piggyback documentation note here at the same time to make them aware that they may have set something up under previous limitations and could take the opportunity now to change it if they want to make changes anyway. I think I have a mixed opinion:

1. I didn't read the old or new wording that way, so I'm wondering if this might cause more confusion, because direction is a separate issue from name vs label. So there's two concepts, each mindbending on their own, in the same sentence.
2. This check is only going to appear for setups where they have a different label, and only mention those specific roles.
3. Making that type of change might also involve some staff retraining for them, which doesn't really have anything to do with this PR which only requires sysadmin involvement and only if they want to futureproof.

Perhaps instead at the end of the message we could include a link to some TBD docs? `<a href="docs link">Learn more</a> about other features available when configuring case roles in xml files.` ?

Reply:
Yeah I think you're right--the parenthetical was an oblique nudge to figure more out; we should just do that explicitly with the docs link.